financial_calculations/coupon_rates.py

```diff
@@ -0,0 +1,65 @@
+import numpy as np
+import pandas as pd
+from datetime import datetime
+from dateutil.relativedelta import relativedelta
+from utils import (
+    get_ZCB_vector
+)
+from financial_calculations.bond_cash_flows import (
+    PMTS_PER_YEAR
+)
+
+def calculate_coupon_rate(start_date, maturity_years, par_value, forward_curve):
+    """
+    Calculate the coupon rate required to produce a par price for a bond.
+    
+    Parameters:
+    -----------
+    start_date : datetime or datetime64[D]
+        The start date of the bond.
+    maturity_years : int
+        The maturity years of the bond.
+    par_value : float
+        The par value of the bond.
+    forward_curve : tuple
+        A tuple containing two elements:
+        - spot_rate_dates (np.ndarray): Array of spot rate dates.
+        - spot_rates (np.ndarray): Array of spot rates corresponding to each spot rate date.
+
+    Returns:
+    --------
+    float:
+        The coupon rate required to produce a par price for the bond.
+    
+    Raises:
+    -------
+    ValueError:
+        If the start date is before the market close date.
+    """
+    spot_rate_dates, spot_rates = forward_curve
+    market_close_date = spot_rate_dates[0]
+
+    # If the market close date is input as a datetime64[D] type, convert to datetime to ensure compatibility with the forward curve data
+    if isinstance(market_close_date, np.datetime64): 
```

Comment:
I wonder if this check is already happening in the conversion code.  maybe kill the if and just call the conversion.  we can try timing it